### PR TITLE
Expose `PyDict_GetItemWithError` on `PyDict` object

### LINKED ISSUE #565: `PyDict` should expose a wrapper over `PyDict_GetItemWithError`
Currently `PyDict::get_item` wraps `PyDict_GetItem` and returns an `Option::None` when the key is missing or an exception occurs. It makes code incapable to differentiate between the two scenarios and code like this likely incorrect:

```
dict.get_item("key").ok_or_else(|| KeyError::py_err("key"))?;
```

### PR DESCRIPTION
**Pull request by dswij**
Closes #565

Please consider adding the following to your pull request:
 - [x] an entry in CHANGELOG.md
 - [x] docs to all new functions and / or detail in the guide
 - [x] tests for all new or changed functions

### CODE REVIEW

File: src/types/dict.rs [hunk 1]
```diff
@@ -562,6 +583,32 @@ mod tests {
...
+                dict.get_item_with_error(7i32)
+                    .unwrap()
+                    .unwrap()
+                    .extract::<i32>()
+                    .unwrap()
+            );
+            assert!(dict.get_item_with_error(8i32).unwrap().is_none());
+
+            let dict_clone = dict.clone();
+            if let Err(err) = dict.get_item_with_error(dict_clone) {
```
Comment (davidhewitt) [inline src/types/dict.rs:604]: Nit: use `.unwrap_err()` here.

**Review by davidhewitt — COMMENTED** [review]
Thanks - implementation looks fine to me! Please add an `## Added ` entry in the CHANGELOG.

File: src/types/dict.rs (hunk 1 above)
Comment (dswij) [reply src/types/dict.rs]: Done [3f1df62](https://github.com/PyO3/pyo3/pull/2536/commits/3f1df62557a926b8d7b54e6f101319d806c2310c)

**Review by messense — APPROVED** [review]
👍

I'm trying to fix the Windows coverage issue in https://github.com/taiki-e/cargo-llvm-cov/pull/203

**Pull request merged**